### PR TITLE
Have public save and load methods, such that they can be used by Babelfish

### PR DESCRIPTION
**Pull request by roesslerj**
Only a refactoring, shouldn't change impl details.

### CODE REVIEW

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 1]
```diff
@@ -102,12 +103,27 @@ public void save( final URI identifier, final T element ) throws IOException {
...
 		final Path path = Paths.get( identifier );
+		try {
+			return load( newInputStream( path ), identifier );
+		} catch ( final IncompatibleReportVersionException | NoSuchFileException e ) {
+			throw e;
+		}
+	}
+
+	@SuppressWarnings( "unchecked" )
+	public T load( final InputStream in, final URI identifier ) throws IOException {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java:120]: Do you have `URI` with your stream? If not, I would try to remove the identifier (as it is only used in the exception). I would generally try to keep the methods the same:
```
T load( final URI identifier ) throws IOException;
T load( final InputStream identifier ) throws IOException;
```

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 2]
```diff
@@ -88,11 +90,10 @@ public void save( final URI identifier, final T element ) throws IOException {
 		final Path path = Paths.get( identifier );
 		final File file = path.toFile();
 		FileUtil.ensureFolder( path.toFile() );
-		try ( final Output output = new Output( new LZ4FrameOutputStream( newOutputStream( path ) ) ) ) {
-			output.writeString( version );
+		try ( OutputStream outputStream = newOutputStream( path ) ) {
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: Please use try-with-resources here, just pass in the OutputStream (similarly to `load`). You should always close the outer most stream only, which should take care of dependent streams.
```
try {
    return save( newOutputStream( path ), element );
} catch ( final Throwable t ) {
    ....
}
```

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 3]
```diff
@@ -102,12 +103,27 @@ public void save( final URI identifier, final T element ) throws IOException {
 		}
 	}
 
-	@SuppressWarnings( "unchecked" )
+	public void save( final OutputStream outputStream, final T element ) throws IOException {
+		final Output output = new Output( new LZ4FrameOutputStream( outputStream ) );
+		output.writeString( version );
+		kryo.writeClassAndObject( output, element );
+		output.close();
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: This output will not be closed on error, see above.

**Review by diba1013 — CHANGES_REQUESTED** [review]

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java (hunk 1 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java:120]: Yeah, I tried that ... but I think this is really helpful information in case of errors. So how could you remove that then?

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java (hunk 3 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: On error, this is no problem, since the resulting file would be corrupt anyways. This close is not there to prevent memory leaks (which most often is the result of forgetting to call close), but to prevent a corrupt result file.

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java (hunk 2 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: Hmm. I am sorry. Maybe I do not understand correctly. But this currently is try-with-resource, right?

Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: Closing a stream multiple times is no problem. Closing an outer stream will also close an inner stream, but closing an inner stream does not close the outer stream. Therefore, close twice ... just to make sure.

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 4]
```diff
@@ -102,12 +103,35 @@ public void save( final URI identifier, final T element ) throws IOException {
...
+
+	/**
+	 * You need to maintain the stream and are responsible for closing it properly.
+	 */
+	@SuppressWarnings( "unchecked" )
+	public T load( final InputStream in, final URI identifier ) throws IOException {
 		String writerVersion = null;
-		try ( final Input input = new Input( new LZ4FrameInputStream( newInputStream( path ) ) ) ) {
+		try {
+			final Input input = new Input( new LZ4FrameInputStream( in ) );
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: Why did you change this? This was fine as it is.

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 5]
```diff
@@ -102,12 +103,35 @@ public void save( final URI identifier, final T element ) throws IOException {
...
-	@SuppressWarnings( "unchecked" )
+	/**
+	 * You need to maintain the stream and are responsible for closing it properly.
+	 */
+	public void save( final OutputStream outputStream, final T element ) throws IOException {
+		final Output output = new Output( new LZ4FrameOutputStream( outputStream ) );
+		output.writeString( version );
+		kryo.writeClassAndObject( output, element );
+		// Flush the stream to avoid corrupt file data
+		output.flush();
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: I thought we discussed this as try-with-resources? Just to close the outer stream at all times.
```
try (final Output output = new Output( new LZ4FrameOutputStream( outputStream ) ) ) {
    output.writeString( version );
    kryo.writeClassAndObject( output, element );
}
```

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 6]
```diff
@@ -102,12 +103,35 @@ public void save( final URI identifier, final T element ) throws IOException {
...
+	}
+
 	@Override
 	public T load( final URI identifier ) throws IOException {
 		final Path path = Paths.get( identifier );
+		try ( InputStream input = newInputStream( path ) ) {
+			return load( input, identifier );
+		} catch ( final IncompatibleReportVersionException | NoSuchFileException e ) {
+			throw e;
+		}
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: We can just let the exception fall through:
```
try ( InputStream input = newInputStream( path ) ) {
	return load( input, identifier );
}
```

**Review by diba1013 — CHANGES_REQUESTED** [review]

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java (hunk 4 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: No it was not. We _must_ close the outermost stream, not some stream "in between". Closing here would have done the latter...

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 7]
```diff
@@ -102,12 +103,33 @@ public void save( final URI identifier, final T element ) throws IOException {
 		}
 	}
 
-	@SuppressWarnings( "unchecked" )
+	/**
+	 * You need to maintain the stream and are responsible for closing it properly.
+	 */
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: This comment is not true anymore

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java [hunk 8]
```diff
@@ -102,12 +103,33 @@ public void save( final URI identifier, final T element ) throws IOException {
...
+		try ( InputStream input = newInputStream( path ) ) {
+			return load( input, identifier );
+		} catch ( final IncompatibleReportVersionException | NoSuchFileException e ) {
+			throw e;
+		}
+	}
+
+	/**
+	 * You need to maintain the stream and are responsible for closing it properly.
+	 */
```
Comment (diba1013) [inline src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: This comment is not true anymore

**Review by diba1013 — CHANGES_REQUESTED** [review]

File: src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java (hunk 7 above)
Comment (roesslerj) [reply src/main/java/de/retest/recheck/persistence/bin/KryoPersistence.java]: Changes were lost during rebase. How annoying. Readded them.

**Review by diba1013 — APPROVED** [review]
